src/ui/ui_settings/ui_settings_service.js

```diff
@@ -0,0 +1,130 @@
+import { defaultsDeep, noop } from 'lodash';
+import { errors as esErrors } from 'elasticsearch';
+
+import { getDefaultSettings } from './defaults';
+
+function hydrateUserSettings(userSettings) {
+  return Object.keys(userSettings)
+    .map(key => ({ key, userValue: userSettings[key] }))
+    .filter(({ userValue }) => userValue !== null)
+    .reduce((acc, { key, userValue }) => ({ ...acc, [key]: { userValue } }), {});
+}
+
+export class UiSettingsService {
+  constructor(options = {}) {
+    const {
+      index,
+      type,
+      id,
+      callCluster,
+      readInterceptor = noop,
+    } = options;
+
+    this._callCluster = callCluster;
+    this._readInterceptor = readInterceptor;
+    this._index = index;
+    this._type = type;
+    this._id = id;
+  }
+
+  getDefaults() {
+    return getDefaultSettings();
+  }
+
+  // returns a Promise for the value of the requested setting
+  async get(key) {
+    const all = await this.getAll();
+    return all[key];
+  }
+
+  async getAll() {
+    const raw = await this.getRaw();
+
+    return Object.keys(raw)
+      .reduce((all, key) => {
+        const item = raw[key];
+        const hasUserValue = 'userValue' in item;
+        all[key] = hasUserValue ? item.userValue : item.value;
+        return all;
+      }, {});
+  }
+
+  async getRaw() {
+    const userProvided = await this.getUserProvided();
+    return defaultsDeep(userProvided, this.getDefaults());
+  }
+
+  async getUserProvided(options) {
+    return hydrateUserSettings(await this._read(options));
+  }
+
+  async setMany(changes) {
+    await this._write(changes);
+  }
+
+  async set(key, value) {
+    await this.setMany({ [key]: value });
+  }
+
+  async remove(key) {
+    await this.set(key, null);
+  }
+
+  async removeMany(keys) {
+    const changes = {};
+    keys.forEach(key => {
+      changes[key] = null;
+    });
+    await this.setMany(changes);
+  }
+
+  async _write(changes) {
+    await this._callCluster('update', {
+      index: this._index,
+      type: this._type,
+      id: this._id,
+      body: {
+        doc: changes
+      }
+    });
+  }
+
+  async _read(options = {}) {
+    const interceptValue = await this._readInterceptor(options);
```

Comment:
We do it we want `_readInterceptor` to support async logic, which it might need to do if the interceptor implemented some sort of auth check, or read values from a different cluster. IMO, using async for "extensions" when possible just make the extensions more usable.